tests/module_api/test_api.py

```diff
@@ -773,6 +774,49 @@ def test_create_room(self) -> None:
         # Check room alias.
         self.assertIsNone(room_alias)
 
+    def test_on_logged_out(self) -> None:
+        """Test that on_logged_out module hook is properly call when logging out
+        a device, and that related pushers are still available at this time.
+        """
+        device_id = "AAAAAAA"
+        user_id = self.register_user("test_on_logged_out", "secret")
+        self.login("test_on_logged_out", "secret", device_id)
+
+        self.get_success(
+            self.hs.get_pusherpool().add_or_update_pusher(
+                user_id=user_id,
+                device_id=device_id,
+                kind="http",
+                app_id="m.http",
+                app_display_name="HTTP Push Notifications",
+                device_display_name="pushy push",
+                pushkey="a@example.com",
+                lang=None,
+                data={"url": "http://example.com/_matrix/push/v1/notify"},
+            )
+        )
+
+        self.nb_of_pushers_in_callback: Optional[int] = None
+
+        self.module_api.register_password_auth_provider_callbacks(
+            on_logged_out=self._on_logged_out_mock
+        )
+
+        device_handler = self.hs.get_device_handler()
+        assert isinstance(device_handler, DeviceHandler)
+        self.get_success(device_handler.delete_devices(user_id, [device_id]))
+
+        self.assertEqual(self.nb_of_pushers_in_callback, 1)
+
+        self.assertEqual(len(self.hs.get_pusherpool().pushers[user_id].values()), 0)
+
+    async def _on_logged_out_mock(
+        self, user_id: str, device_id: Optional[str], access_token: str
+    ) -> None:
+        self.nb_of_pushers_in_callback = len(
+            self.hs.get_pusherpool().pushers[user_id].values()
+        )
```

Comment:
I think this is clearer if we don't hang state on the test case object:

```suggestion
        # Setup a callback to count the number of pushers.
        number_of_pushers_in_callback: Optional[int] = None

	    async def on_logged_out_mock(
	        self, user_id: str, device_id: Optional[str], access_token: str
	    ) -> None:
	        nonlocal number_of_pushers_in_callback
	        number_of_pushers_in_callback = len(
	            self.hs.get_pusherpool().pushers[user_id].values()
	        )

        self.module_api.register_password_auth_provider_callbacks(
            on_logged_out=on_logged_out_mock
        )

        # Delete the device.
        device_handler = self.hs.get_device_handler()
        assert isinstance(device_handler, DeviceHandler)
        self.get_success(device_handler.delete_devices(user_id, [device_id]))

        # Check that the callback was called and the pushers still existed.
        self.assertEqual(self.nb_of_pushers_in_callback, 1)

        # Ensure the pushers were deleted after the callback.
        self.assertEqual(len(self.hs.get_pusherpool().pushers[user_id].values()), 0)
```